src/services/telemetryService.ts

```diff
@@ -115,34 +29,167 @@ interface TelemetryOptions {
 
 	userId?: string;
 
-	[prop: string]: any;
-}
-
-export interface UserFlowTelemetry extends TelemetryOptions {
-	flow: string;
+	[key: string]: any;
 }
 
 export interface ITelemetryService {
-	sendTelemetryEvent(gitpodHost: string, eventName: string, properties?: TelemetryEventProperties): void;
-	sendRawTelemetryEvent(gitpodHost: string, eventName: string, properties?: RawTelemetryEventProperties): void;
-	sendTelemetryException(gitpodHost: string, error: Error, properties?: TelemetryEventProperties): void;
+	sendTelemetryEvent(eventName: string, properties?: TelemetryEventProperties): void;
+	sendTelemetryException(error: Error, properties?: TelemetryEventProperties): void;
 
-	sendUserFlowStatus(status: string, flow: UserFlowTelemetry): void;
+	sendUserFlowStatus(status: string, flow: UserFlowTelemetryProperties): void;
 }
 
-export class TelemetryService extends BaseTelemetryReporter implements ITelemetryService {
-	constructor(extensionId: string, extensionVersion: string, segmentKey: string, logger: vscode.LogOutputChannel) {
-		super(extensionId, extensionVersion, {
-			release: os.release(),
-			platform: os.platform(),
-			architecture: os.arch(),
-		}, gitpodHost => analyticsClientFactory(gitpodHost, segmentKey, logger));
+const TRUSTED_VALUES = new Set([
+	'gitpodHost'
+]);
+
+export class TelemetryService extends Disposable implements ITelemetryService {
+	private analitycsClients: Map<string, Analytics> = new Map();
+	private telemetryLogger: vscode.TelemetryLogger;
+
+	constructor(segmentKey: string, private readonly logService: ILogService) {
+		super();
+
+		this.telemetryLogger = this._register(vscode.env.createTelemetryLogger(
+			{
+				sendEventData: (eventName, data) => {
+					const idx = eventName.indexOf('/');
+					eventName = eventName.substring(idx + 1);
+
+					const properties = data ?? {};
+
+					const gitpodHost: string | undefined = properties['gitpodHost'];
+					if (!gitpodHost) {
+						logService.error(`Missing 'gitpodHost' property in event ${eventName}`);
+						return;
+					}
+
+					delete properties['gitpodHost'];
+
+					this.getSegmentAnalyticsClient(gitpodHost, segmentKey)?.track({
+						anonymousId: vscode.env.machineId,
+						event: eventName,
+						properties
+					}, (err) => {
+						if (err) {
+							logService.error('Failed to log event to app analytics:', err);
+						}
+					});
+				},
+				sendErrorData: (error, data) => {
+					const properties = data ?? {};
+
+					// Unhandled errors have no data so use host from config
+					const gitpodHost = properties['gitpodHost'] ?? Configuration.getGitpodHost();
```

Comment:
gitpodHost is vscode.TelemetryTrustedValue

Reply:
What's the benefit to use `vscode.TelemetryTrustedValue`?

Reply:
Ah I see it's only unwrapped for logEvent not for error, [there's a TODO](https://github.com/microsoft/vscode/blob/44ef5cc53127cbaa11dee1728bdf8c24522f8fa0/src/vs/workbench/api/common/extHostTelemetry.ts#L278-L279) about it :facepalm:

`vscode.TelemetryTrustedValue` means it won't be anonymized